packages/python/src/components/FunctionDeclaration.tsx

```diff
@@ -1,15 +1,65 @@
-import { emitSymbol, Name, Show } from "@alloy-js/core";
+import {
+  Children,
+  childrenArray,
+  code,
+  DeclarationContext,
+  emitSymbol,
+  findKeyedChild,
+  findUnkeyedChildren,
+  List,
+  Name,
+  Refkey,
+  Show,
+  taggedComponent,
+  useContext,
+} from "@alloy-js/core";
+import { abcModule } from "../builtins/python.js";
+import { PythonOutputSymbol } from "../index.js";
+import { ParameterDescriptor } from "../parameter-descriptor.js";
 import { createPythonSymbol } from "../symbol-creation.js";
+import { usePythonScope } from "../symbols/scopes.js";
 import { getCallSignatureProps } from "../utils.js";
 import { CallSignature, CallSignatureProps } from "./CallSignature.jsx";
 import { BaseDeclarationProps, Declaration } from "./Declaration.js";
 import { PythonBlock } from "./PythonBlock.jsx";
-import { LexicalScope, NoNamePolicy } from "./index.js";
+import {
+  Atom,
+  LexicalScope,
+  NoNamePolicy,
+  TypeExpressionProps,
+} from "./index.js";
+
+const setterTag = Symbol();
+const deleterTag = Symbol();
+
+/**
+ * Validates that the current scope is a member scope (inside a class).
+ * Throws an error if not in a member scope.
+ */
+function validateMemberScope(name: string, type: string = "Method") {
+  const currentScope = usePythonScope();
+  if (!currentScope?.isMemberScope) {
+    throw new Error(
+      `${type} "${name}" must be declared inside a class (member scope)`,
+    );
+  }
+}
 
 export interface FunctionDeclarationProps
   extends BaseDeclarationProps,
     CallSignatureProps {
+  /**
+   * Indicates that the function is async.
+   */
   async?: boolean;
+  /**
+   * Indicates the type of function.
+   */
+  functionType?: "instance" | "class" | "static";
+  /**
+   * Optional existing symbol to use instead of creating a new one.
+   */
+  sym?: PythonOutputSymbol;
```

Comment:
We're still exposing these as part of the "API" of `FunctionDeclaration`. Is that intentional? IMO the way to solve this is to have a `BaseFunctionDeclaration` (which is _not_ exported) that has these properties, and extend it to define `FunctionDeclaration` without them.

I see you referenced a `FunctionDeclarationBase` in [one of your comments](https://github.com/pinterest/alloy/pull/7/files#r2337812287) … what happened to that? Sounds like what I am describing here.